src/Compilers/Core/Portable/Compilation/IExpression.cs

```diff
@@ -0,0 +1,778 @@
+        EnumPrefixIncrement,
+        EnumPrefixDecrement,
+
+        PointerPostfixIncrement,
+        PointerPostfixDecrement,
+        PointerPrefixIncrement,
+        PointerPrefixDecrement,
+
+        DynamicBitwiseNegation,
+        DynamicLogicalNot,
+        DynamicTrue,
+        DynamicFalse,
+        DynamicPlus,
+        DynamicMinus,
+        DynamicPostfixIncrement,
+        DynamicPostfixDecrement,
+        DynamicPrefixIncrement,
+        DynamicPrefixDecrement,
+
+        ObjectPlus,
+        ObjectMinus,
+        ObjectNot
+    }
+
+    /// <summary>
+    /// Represents an operation with two operands that produces a result with the same type as at least one of the operands.
+    /// </summary>
+    public interface IBinary : IWithOperator
+    {
+        BinaryOperationKind BinaryKind { get; }
+        IExpression Left { get; }
+        IExpression Right { get; }
+    }
+
+    /// <summary>
+    /// Kinds of binary operations.
+    /// </summary>
+    public enum BinaryOperationKind
+    {
+        None,
+
+        OperatorAdd,
+        OperatorSubtract,
+        OperatorMultiply,
+        OperatorDivide,
+        OperatorRemainder,
+        OperatorLeftShift,
+        OperatorRightShift,
+        OperatorAnd,
+        OperatorOr,
+        OperatorXor,
+        OperatorConditionalAnd,
+        OperatorConditionalOr,
+
+        IntegerAdd,
+        IntegerSubtract,
+        IntegerMultiply,
+        IntegerDivide,
+        IntegerRemainder,
+        IntegerLeftShift,
+        IntegerRightShift,
+        IntegerAnd,
+        IntegerOr,
+        IntegerXor,
+
+        UnsignedAdd,
+        UnsignedSubtract,
+        UnsignedMultiply,
+        UnsignedDivide,
+        UnsignedRemainder,
+        UnsignedLeftShift,
+        UnsignedRightShift,
+        UnsignedAnd,
+        UnsignedOr,
+        UnsignedXor,
+
+        FloatingAdd,
+        FloatingSubtract,
+        FloatingMultiply,
+        FloatingDivide,
+        FloatingRemainder,
+        FloatingPower,
+
+        DecimalAdd,
+        DecimalSubtract,
+        DecimalMultiply,
+        DecimalDivide,
+
+        BooleanAnd,
+        BooleanOr,
+        BooleanXor,
+        BooleanConditionalAnd,
+        BooleanConditionalOr,
+
+        EnumAdd,
+        EnumSubtract,
+        EnumAnd,
+        EnumOr,
+        EnumXor,
+
+        PointerIntegerAdd,
+        IntegerPointerAdd,
+        PointerIntegerSubtract,
+        PointerSubtract,
+
+        DynamicAdd,
+        DynamicSubtract,
+        DynamicMultiply,
+        DynamicDivide,
+        DynamicRemainder,
+        DynamicLeftShift,
+        DynamicRightShift,
+        DynamicAnd,
+        DynamicOr,
+        DynamicXor,
+
+        ObjectAdd,
+        ObjectSubtract,
+        ObjectMultiply,
+        ObjectDivide,
+        ObjectPower,
+        ObjectIntegerDivide,
+        ObjectRemainder,
+        ObjectLeftShift,
+        ObjectRightShift,
+        ObjectAnd,
+        ObjectOr,
+        ObjectXor,
```

Comment:
More naming.  Is Roslyn naming policy to use things like "Xor" or to be more explicit with something like "ExclusiveOr"?